File: sdk/typescript/src/validator.ts
import { assertJSON, ContractError, fail, type JSONValue } from "./json.js";
import { schemas } from "./schema-data.js";
import {
  object,
  validateAgainst,
  validateSchema,
  type ObjectValue,
} from "./schema.js";
import { pointerParts, reference } from "./mapping.js";
export interface ValidationError {
  code: string;
  message: string;
  path?: string;
}
export interface ValidationResult {
  valid: boolean;
  errors: ValidationError[];
}
const check = (fn: () => void): ValidationResult => {
  try {
    fn();
    return { valid: true, errors: [] };
  } catch (e) {
    if (e instanceof ContractError)
      return { valid: false, errors: [{ code: e.code, message: e.message }] };
    throw e;
  }
};
export function validateTask(task: JSONValue): void {
  assertJSON(task);
  if (!validateAgainst(task, schemas["task.schema.json"])) fail("INVALID_TASK");
  const t = object(task);
  validateSchema(t.inputSchema, schemas["payload-schema.schema.json"]);
  validateSchema(t.outputSchema, schemas["payload-schema.schema.json"]);
  const retry = object(t.retry ?? {});
  if (Number(retry.initialDelayMs ?? 1000) > Number(retry.maxDelayMs ?? 30000))
    fail("INVALID_TASK");
  if (
    t.recovery === "idempotent" &&
    Number(t.idempotencyWindowMs) < 5000 + Number(t.timeoutMs ?? 300000)
  )
    fail("INVALID_TASK");
}
export function normalizeTask(task: JSONValue): ObjectValue {
  assertJSON(task);
  validateTask(task);
  const t = object(task);
  return {
    ...t,
    timeoutMs: t.timeoutMs ?? 300000,
    retry: {
      maxAttempts: 3,
      initialDelayMs: 1000,
      maxDelayMs: 30000,
      ...object(t.retry ?? {}),
    },
  };
}
function guaranteed(schema: JSONValue, parts: string[]): boolean {
  if (!parts.length) return true;
  const s = object(schema);
  if (s.oneOf)
    return (s.oneOf as JSONValue[]).every((x) => guaranteed(x, parts));
  const [key, ...rest] = parts;
  if (s.type === "object") {
    const p = object(s.properties ?? {});
    return (
      Array.isArray(s.required) &&
      s.required.includes(key) &&
      Object.hasOwn(p, key) &&
      guaranteed(p[key], rest)
    );
  }
  if (s.type === "array")
    return (
      /^(0|[1-9][0-9]*)$/.test(key) &&
      key === String(Number(key)) &&
      Number(key) < Number(s.minItems ?? 0) &&
      !!s.items &&
      guaranteed(s.items, rest)
    );
  return false;
}
function workflow(manifest: JSONValue, definitions: JSONValue[]): void {
  const m = object(manifest);
  if (m.manifestVersion !== 1) fail("UNSUPPORTED_MANIFEST_VERSION");
  if (!Array.isArray(m.nodes) || m.nodes.length === 0) fail("EMPTY_NODES");
  if (m.nodes.length > 50) fail("NODE_COUNT_EXCEEDED");
  if (!validateAgainst(m, schemas["workflow.schema.json"]))
    fail("INVALID_MANIFEST");
  validateSchema(m.inputSchema, schemas["payload-schema.schema.json"]);
  validateSchema(m.outputSchema, schemas["payload-schema.schema.json"]);
  const tasks = new Map<string, ObjectValue>();
  for (const t of definitions) {
    validateTask(t);
    const task = object(t);
    if (tasks.has(String(task.name))) fail("INVALID_TASK");
    tasks.set(String(task.name), task);
  }
  const nodes = m.nodes as ObjectValue[],
    byId = new Map<string, ObjectValue>();
  for (const n of nodes) {
    const id = String(n.id);
    if (byId.has(id)) fail("DUPLICATE_NODE_ID");
    byId.set(id, n);
    if (n.type !== "task") fail("UNSUPPORTED_CAPABILITY");
    if (!tasks.has(String(n.task))) fail("MISSING_TASK_REF");
  }
  for (const n of nodes)
    for (const d of (n.after ?? []) as string[]) {
      if (!byId.has(d)) fail("MISSING_DEPENDENCY");
    }
  const ancestors = new Map<string, Set<string>>(),
    visiting = new Set<string>();
  const visit = (id: string): Set<string> => {
    if (visiting.has(id)) fail("CYCLE_DETECTED");
    if (ancestors.has(id)) return ancestors.get(id)!;
    visiting.add(id);
    const set = new Set<string>();
    for (const d of (byId.get(id)!.after ?? []) as string[]) {
      set.add(d);
      for (const x of visit(d)) set.add(x);
    }
    visiting.delete(id);
    ancestors.set(id, set);
    return set;
  };
  nodes.forEach((n) => visit(String(n.id)));
  const used = new Set<string>();
  const mapping = (
    v: JSONValue,
    allowed: Set<string>,
    isOutput = false,
  ): void => {
    if (v === null || typeof v !== "object") return;
    if (Array.isArray(v)) {
      v.forEach((x) => mapping(x, allowed, isOutput));
      return;
    }
    if (Object.hasOwn(v, "literal")) {
      if (Object.keys(v).length !== 1) fail("INPUT_MAPPING_ERROR");
      return;
    }
    if (Object.hasOwn(v, "$ref")) {
      reference(v);
      let source = m.inputSchema;
      if (v.$ref === "step.output") {
        const id = String(v.stepId);
        if (!allowed.has(id)) fail("INPUT_MAPPING_ERROR");
        source = tasks.get(String(byId.get(id)!.task))!.outputSchema;
        if (isOutput) used.add(id);
      }
      if (
        !guaranteed(source, pointerParts(v.pointer)) &&
        !Object.hasOwn(v, "default")
      )
        fail("INPUT_MAPPING_ERROR");
      return;
    }
    Object.values(v).forEach((x) => mapping(x, allowed, isOutput));
  };
  nodes.forEach((n) => mapping(n.input ?? {}, ancestors.get(String(n.id))!));
  mapping(m.output, new Set(byId.keys()), true);
  const referencedAsDependency = new Set(
    nodes.flatMap((n) => (n.after ?? []) as string[]),
  );
  for (const n of nodes)
    if (
      !referencedAsDependency.has(String(n.id)) &&
      !used.has(String(n.id)) &&
      n.sideEffect !== true
    )
      fail("ORPHAN_LEAF");
}
export function validateWorkflowManifest(
  manifest: unknown,
  tasks: unknown = [],
): ValidationResult {
  return check(() => {
    assertJSON(manifest);
    assertJSON(tasks);
    if (!Array.isArray(tasks)) fail("INVALID_TASK");
    workflow(manifest, tasks);
  });
}
export function validateDeployment(value: unknown): ValidationResult {
  return check(() => {
    assertJSON(value);
    if (!validateAgainst(value, schemas["deployment.schema.json"]))
      fail("INVALID_MANIFEST");
    const m = object(value),
      names = new Set<string>();
    for (const w of m.workflows as JSONValue[]) {
      const name = String(object(w).name);
      if (names.has(name)) fail("INVALID_MANIFEST");
      names.add(name);
      workflow(w, m.tasks as JSONValue[]);
    }
  });
}

File: internal/contracts/validator.go
package contracts

import "strconv"

func ValidateTask(task any) error {
	if err := checkJSON(task, 0); err != nil {
		return err
	}
	if !ValidateAgainst(task, "manifest/task.schema.json") {
		return failure("INVALID_TASK")
	}
	t := obj(task)
	for _, key := range []string{"inputSchema", "outputSchema"} {
		if err := ValidateSchema(t[key]); err != nil {
			return err
		}
	}
	retry := obj(t["retry"])
	if number(retry["initialDelayMs"], 1000) > number(retry["maxDelayMs"], 30000) {
		return failure("INVALID_TASK")
	}
	if t["recovery"] == "idempotent" && number(t["idempotencyWindowMs"], 0) < 5000+number(t["timeoutMs"], 300000) {
		return failure("INVALID_TASK")
	}
	return nil
}
func NormalizeTask(task any) (any, error) {
	if err := checkJSON(task, 0); err != nil {
		return nil, err
	}
	if err := ValidateTask(task); err != nil {
		return nil, err
	}
	out := map[string]any{}
	for k, v := range obj(task) {
		out[k] = v
	}
	out["timeoutMs"] = number(out["timeoutMs"], 300000)
	retry := map[string]any{"maxAttempts": float64(3), "initialDelayMs": float64(1000), "maxDelayMs": float64(30000)}
	for k, v := range obj(out["retry"]) {
		retry[k] = v
	}
	out["retry"] = retry
	return out, nil
}
func guaranteed(schema any, parts []string) bool {
	if len(parts) == 0 {
		return true
	}
	s := obj(schema)
	if variants := arr(s["oneOf"]); variants != nil {
		for _, v := range variants {
			if !guaranteed(v, parts) {
				return false
			}
		}
		return true
	}
	key := parts[0]
	if s["type"] == "object" {
		p := obj(s["properties"])
		value, ok := p[key]
		return ok && contains(arr(s["required"]), key) && guaranteed(value, parts[1:])
	}
	if s["type"] == "array" {
		i, err := strconv.Atoi(key)
		return err == nil && indexPattern.MatchString(key) && float64(i) < number(s["minItems"], 0) && s["items"] != nil && guaranteed(s["items"], parts[1:])
	}
	return false
}
func ValidateWorkflow(manifest any, definitions []any) error {
	if err := checkJSON(manifest, 0); err != nil {
		return err
	}
	m := obj(manifest)
	if m == nil {
		return failure("INVALID_MANIFEST")
	}
	if m["manifestVersion"] != float64(1) {
		return failure("UNSUPPORTED_MANIFEST_VERSION")
	}
	nodes := arr(m["nodes"])
	if len(nodes) == 0 {
		return failure("EMPTY_NODES")
	}
	if len(nodes) > 50 {
		return failure("NODE_COUNT_EXCEEDED")
	}
	if !ValidateAgainst(m, "manifest/workflow.schema.json") {
		return failure("INVALID_MANIFEST")
	}
	for _, key := range []string{"inputSchema", "outputSchema"} {
		if err := ValidateSchema(m[key]); err != nil {
			return err
		}
	}
	tasks := map[string]map[string]any{}
	for _, t := range definitions {
		if err := ValidateTask(t); err != nil {
			return err
		}
		task := obj(t)
		name := str(task["name"])
		if tasks[name] != nil {
			return failure("INVALID_TASK")
		}
		tasks[name] = task
	}
	byID := map[string]map[string]any{}
	for _, v := range nodes {
		n := obj(v)
		id := str(n["id"])
		if byID[id] != nil {
			return failure("DUPLICATE_NODE_ID")
		}
		byID[id] = n
		if n["type"] != "task" {
			return failure("UNSUPPORTED_CAPABILITY")
		}
		if tasks[str(n["task"])] == nil {
			return failure("MISSING_TASK_REF")
		}
	}
	for _, v := range nodes {
		for _, dep := range arr(obj(v)["after"]) {
			d := str(dep)
			if byID[d] == nil {
				return failure("MISSING_DEPENDENCY")
			}
		}
	}
	ancestors := map[string]map[string]bool{}
	visiting := map[string]bool{}
	var visit func(string) (map[string]bool, error)
	visit = func(id string) (map[string]bool, error) {
		if visiting[id] {
			return nil, failure("CYCLE_DETECTED")
		}
		if a := ancestors[id]; a != nil {
			return a, nil
		}
		visiting[id] = true
		set := map[string]bool{}
		for _, dep := range arr(byID[id]["after"]) {
			d := str(dep)
			set[d] = true
			other, err := visit(d)
			if err != nil {
				return nil, err
			}
			for x := range other {
				set[x] = true
			}
		}
		delete(visiting, id)
		ancestors[id] = set
		return set, nil
	}
	for _, v := range nodes {
		if _, err := visit(str(obj(v)["id"])); err != nil {
			return err
		}
	}
	used := map[string]bool{}
	var mapping func(any, map[string]bool, bool) error
	mapping = func(value any, allowed map[string]bool, isOutput bool) error {
		switch v := value.(type) {
		case []any:
			for _, x := range v {
				if err := mapping(x, allowed, isOutput); err != nil {
					return err
				}
			}
		case map[string]any:
			if _, ok := v["literal"]; ok {
				if len(v) != 1 {
					return failure("INPUT_MAPPING_ERROR")
				}
				return nil
			}
			if _, ok := v["$ref"]; ok {
				if err := Reference(v); err != nil {
					return err
				}
				source := m["inputSchema"]
				if v["$ref"] == "step.output" {
					id := str(v["stepId"])
					if !allowed[id] {
						return failure("INPUT_MAPPING_ERROR")
					}
					source = tasks[str(byID[id]["task"])]["outputSchema"]
					if isOutput {
						used[id] = true
					}
				}
				parts, _ := PointerParts(v["pointer"])
				_, hasDefault := v["default"]
				if !guaranteed(source, parts) && !hasDefault {
					return failure("INPUT_MAPPING_ERROR")
				}
				return nil
			}
			for _, x := range v {
				if err := mapping(x, allowed, isOutput); err != nil {
					return err
				}
			}
		}
		return nil
	}
	all := map[string]bool{}
	for _, v := range nodes {
		n := obj(v)
		id := str(n["id"])
		all[id] = true
		if err := mapping(n["input"], ancestors[id], false); err != nil {
			return err
		}
	}
	if err := mapping(m["output"], all, true); err != nil {
		return err
	}
	referencedAsDependency := map[string]bool{}
	for _, v := range nodes {
		for _, dep := range arr(obj(v)["after"]) {
			referencedAsDependency[str(dep)] = true
		}
	}
	for _, v := range nodes {
		n := obj(v)
		id := str(n["id"])
		if !referencedAsDependency[id] && !used[id] && n["sideEffect"] != true {
			return failure("ORPHAN_LEAF")
		}
	}
	return nil
}
func ValidateDeployment(v any) error {
	if err := checkJSON(v, 0); err != nil {
		return err
	}
	if !ValidateAgainst(v, "manifest/deployment.schema.json") {
		return failure("INVALID_MANIFEST")
	}
	m := obj(v)
	names := map[string]bool{}
	for _, w := range arr(m["workflows"]) {
		name := str(obj(w)["name"])
		if names[name] {
			return failure("INVALID_MANIFEST")
		}
		names[name] = true
		if err := ValidateWorkflow(w, arr(m["tasks"])); err != nil {
			return err
		}
	}
	return nil
}
